skimage/filters/rank/core_cy.pyx

```diff
@@ -272,3 +284,239 @@ cdef void _core(void kernel(dtype_t_out*, Py_ssize_t, Py_ssize_t*, double,
+    use kernel function return value for output image.
+    """
+
+    cdef Py_ssize_t planes = image.shape[0]
+    cdef Py_ssize_t rows = image.shape[1]
+    cdef Py_ssize_t cols = image.shape[2]
+    cdef Py_ssize_t splanes = selem.shape[0]
+    cdef Py_ssize_t srows = selem.shape[1]
+    cdef Py_ssize_t scols = selem.shape[2]
+    cdef Py_ssize_t odepth = out.shape[3]
+
+    cdef Py_ssize_t centre_p = (selem.shape[0] // 2) + shift_x
+    cdef Py_ssize_t centre_r = (selem.shape[1] // 2) + shift_y
+    cdef Py_ssize_t centre_c = (selem.shape[2] // 2) + shift_z
+
+    # check that structuring element center is inside the element bounding box
+    if not 0 <= centre_p < splanes:
+        raise ValueError("half selem + shift_x must be between 0 and selem")
+    if not 0 <= centre_r < srows:
+        raise ValueError("half selem + shift_y must be between 0 and selem")
+    if not 0 <= centre_c < scols:
+        raise ValueError("half selem + shift_z must be between 0 and selem")
+
+    cdef Py_ssize_t mid_bin = n_bins // 2
+
+    # define pointers to the data
+    cdef char* mask_data = &mask[0, 0, 0]
+
+    # define local variable types
+    cdef Py_ssize_t r, c, p, j, rr, cc, pp, value, local_max, i, even_row
+
+    # number of pixels actually inside the neighborhood (double)
+    cdef double pop = 0
+
+    # build attack and release borders by using difference along axis
+    t = np.dstack((selem, np.zeros((selem.shape[0], selem.shape[1], 1))))
+    cdef unsigned char[:, :, :] t_e = (np.diff(t, axis=2) < 0).view(np.uint8)
+
+    t = np.dstack((np.zeros((selem.shape[0], selem.shape[1], 1)), selem))
+    cdef unsigned char[:, :, :] t_w = (np.diff(t, axis=2) > 0).view(np.uint8)
+
+    t = np.hstack((selem, np.zeros((selem.shape[0], 1, selem.shape[2]))))
+    cdef unsigned char[:, :, :] t_s = (np.diff(t, axis=1) < 0).view(np.uint8)
+
+    t = np.hstack((np.zeros((selem.shape[0], 1, selem.shape[2])), selem))
+    cdef unsigned char[:, :, :] t_n = (np.diff(t, axis=1) > 0).view(np.uint8)
+
+    # the current local histogram distribution
+    # cdef Py_ssize_t* histo
+    cdef Py_ssize_t* histo
+
+    # these lists contain the relative pixel row and column for each of the 4
+    # attack borders east, west, north and south e.g. se_e_r lists the rows of
+    # the east structuring element border
+    cdef Py_ssize_t se_size = srows * scols * splanes * sizeof(Py_ssize_t)
+    se_e_r = np.zeros(se_size, dtype=np.intp)
+    se_e_c = np.zeros(se_size, dtype=np.intp)
+    se_e_p = np.zeros(se_size, dtype=np.intp)
+
+    se_w_p = np.zeros(se_size, dtype=np.intp)
+    se_w_r = np.zeros(se_size, dtype=np.intp)
+    se_w_c = np.zeros(se_size, dtype=np.intp)
+
+    se_n_p = np.zeros(se_size, dtype=np.intp)
+    se_n_r = np.zeros(se_size, dtype=np.intp)
+    se_n_c = np.zeros(se_size, dtype=np.intp)
+
+    se_s_p = np.zeros(se_size, dtype=np.intp)
+    se_s_r = np.zeros(se_size, dtype=np.intp)
+    se_s_c = np.zeros(se_size, dtype=np.intp)
```

Comment:
```suggestion
    cdef Py_ssize_t se_size = srows * scols * splanes
    cdef Py_ssize_t [::1] se_e_r = np.empty(se_size, dtype=np.intp)
    cdef Py_ssize_t [::1] se_e_c = np.empty(se_size, dtype=np.intp)
    cdef Py_ssize_t [::1] se_e_p = np.empty(se_size, dtype=np.intp)

    cdef Py_ssize_t [::1] se_w_p = np.empty(se_size, dtype=np.intp)
    cdef Py_ssize_t [::1] se_w_r = np.empty(se_size, dtype=np.intp)
    cdef Py_ssize_t [::1] se_w_c = np.empty(se_size, dtype=np.intp)

    cdef Py_ssize_t [::1] se_n_p = np.empty(se_size, dtype=np.intp)
    cdef Py_ssize_t [::1] se_n_r = np.empty(se_size, dtype=np.intp)
    cdef Py_ssize_t [::1] se_n_c = np.empty(se_size, dtype=np.intp)

    cdef Py_ssize_t [::1] se_s_p = np.empty(se_size, dtype=np.intp)
    cdef Py_ssize_t [::1] se_s_r = np.empty(se_size, dtype=np.intp)
    cdef Py_ssize_t [::1] se_s_c = np.empty(se_size, dtype=np.intp)
```
You can find more doc about numpy usage in cython [here](https://cython.readthedocs.io/en/latest/src/userguide/numpy_tutorial.html#numpy-tutorial) ;)